src/NHibernate/Loader/Loader.cs

```diff
@@ -1829,7 +1831,36 @@ protected IList List(ISessionImplementor session, QueryParameters queryParameter
 
 		internal bool IsCacheable(QueryParameters queryParameters)
 		{
-			return _factory.Settings.IsQueryCacheEnabled && queryParameters.Cacheable;
+			if (Factory.Settings.IsQueryCacheEnabled && queryParameters.Cacheable)
+			{
+				if (QuerySpaces?.Count > 0)
+				{
+					ISet<string> neverCachedEntities = new HashSet<string>();
+
+					foreach (var persister in _factory
```

Comment:
Can we move this heavy logic to the query plan cache(bool SupportsQueryCache) or elsewhere, this check for each query will be costly. If its ok could you point me how to implement it.

In here we could just throw exception If query does not supports query cache and get details of entityNames?